src/socket/raw.rs

```diff
@@ -0,0 +1,272 @@
+        &self.payload[..self.size]
+    }
+
+    fn as_mut<'b>(&'b mut self) -> &'b mut [u8] {
+        &mut self.payload[..self.size]
+    }
+}
+
+impl<'a> Resetable for PacketBuffer<'a> {
+    fn reset(&mut self) {
+        self.ip_addr = IpAddress::Unspecified;
+        self.size = 0;
+    }
+}
+
+/// A raw IP packet ring buffer.
+pub type SocketBuffer<'a, 'b: 'a> = RingBuffer<'a, PacketBuffer<'b>>;
+
+/// A Raw IP socket
+///
+/// A raw socket is bound to a specifict IP protocol and owns
+/// transmit and receive packet buffers.
+#[derive(Debug)]
+pub struct RawSocket<'a, 'b: 'a> {
+    ip_addr: IpAddress,
+    ip_proto: IpProtocol,
+    rx_buffer: SocketBuffer<'a, 'b>,
+    tx_buffer: SocketBuffer<'a, 'b>,
+    debug_id: usize,
+}
+
+impl<'a, 'b> RawSocket<'a, 'b> {
+    /// Create a Raw IP socket with the given buffers, bound to the given IP protocol
+    /// and IP address.
+    pub fn new(
+        rx_buffer: SocketBuffer<'a, 'b>,
+        tx_buffer: SocketBuffer<'a, 'b>,
+        ip_addr: IpAddress,
+        ip_proto: IpProtocol,
+    ) -> Socket<'a, 'b> {
+        Socket::Raw(RawSocket {
+            rx_buffer,
+            tx_buffer,
+            debug_id: 0,
+            ip_proto,
+            ip_addr,
+        })
+    }
+
+    /// Return the debug identifier.
+    pub fn debug_id(&self) -> usize {
+        self.debug_id
+    }
+
+    /// Set the debug identifier.
+    ///
+    /// The debug identifier is a number printed in socket trace messages.
+    /// It could as well be used by the user code.
+    pub fn set_debug_id(&mut self, id: usize) {
+        self.debug_id = id;
+    }
+
+    /// Return the IP protocol the socket is bound to.
+    pub fn ip_proto(&self) -> IpProtocol {
+        self.ip_proto
+    }
+
+    /// Check whether the transmit buffer is full.
+    pub fn can_send(&self) -> bool {
+        !self.tx_buffer.full()
+    }
+
+    /// Check whether the receive buffer is not empty.
+    pub fn can_recv(&self) -> bool {
+        !self.rx_buffer.empty()
+    }
+
+    /// Enqueue a packet to be sent to the given IP address, and return a pointer
+    /// to its payload.
+    ///
+    /// This function returns `Err(())` if the size is greater than what
+    /// the transmit buffer can accomodate.
+    pub fn send(&mut self, size: usize, ip_addr: IpAddress) -> Result<&mut [u8], ()> {
+        let packet_buf = self.tx_buffer.enqueue()?;
+        packet_buf.ip_addr = ip_addr;
+        packet_buf.size = size;
+        net_trace!(
+            "[{}]{}:{}: proto {} buffer to send {} octets",
+            self.debug_id,
+            self.ip_addr,
+            packet_buf.ip_addr,
+            self.ip_proto,
+            packet_buf.size
+        );
+        Ok(&mut packet_buf.as_mut()[..size])
+    }
+
+    /// Enqueue a packet to be sent to the given IP address, and fill it from a slice.
+    ///
+    /// See also [send](#method.send).
+    pub fn send_slice(&mut self, data: &[u8], ip_addr: IpAddress) -> Result<usize, ()> {
```

Comment:
There's no need to accept IpAddress in send, or provide it in recv; it is the responsibility of the client to dissect the packet.